src/v/cloud_storage/segment_meta_cstore.cc

```diff
@@ -1002,7 +1002,98 @@ class segment_meta_cstore::impl
+            auto dereference_op,
+            auto committed_offset_dereference_op,
+            auto prev_op) {
+              if (container.empty()) {
+                  // empty container, can be inserted
+                  return true;
+              }
+              // look for an exact match or whatever comes after it
+              auto candidate = container.lower_bound(sm.base_offset);
+              if (
+                candidate == container.begin()
+                && dereference_op(candidate) != sm.base_offset) {
+                  // not an exact hit, and candidate starts after offset. can be
+                  // inserted
+                  return true;
+              }
+
+              if (
+                candidate != container.end()
+                && dereference_op(candidate) == sm.base_offset) {
+                  // candidate is not end() and it's an exact hit, check
+                  // committed_offset and return true if the new
+                  // committed_offset is greater or equal.
+                  return committed_offset_dereference_op(candidate)
+                         <= sm.committed_offset;
+              }
+
+              // candidate comes after the offset and it's not begin(), check
+              // the prev segment as it might contain offset
+              auto prev_it = prev_op(candidate);
+              // first part of condition is always true (it's not less than sm,
+              // and it's not equal)
+              if (
+                dereference_op(prev_it) < sm.base_offset
+                && committed_offset_dereference_op(prev_it)
+                     >= sm.committed_offset) {
+                  // prev already covers the range from sm
+                  return false;
+              }
+              // no perfect overlap
+              return true;
+          };
+
+        auto& base_col = _col.get_column_cref<segment_meta_ix::base_offset>();
+        auto& committed_col
+          = _col.get_column_cref<segment_meta_ix::committed_offset>();
+        // check _write_buffer first, then cstore. use lambda to adapt both to
+        // the implementation
+        return likely(impl_can_be_inserted(
+                 m,
+                 _write_buffer,
+                 [](auto& it) { return it->second.base_offset; },
+                 [](auto& it) { return it->second.committed_offset; },
+                 [](auto& it) { return std::prev(it); }))
+               && impl_can_be_inserted(
+                 m,
+                 base_col,
+                 [](auto& it) { return *it; },
+                 [&](auto& it) { return *committed_col.at_index(it.index()); },
+                 [&](auto& it) {
+                     return base_col.at_index(
+                       (it.is_end() ? _col.size() : it.index()) - 1);
+                 });
+    }
+    bool insert(const segment_meta& m) {
+        if (unlikely(!can_be_inserted(m))) {
+            return false;
+        }
```

Comment:
this looks like it belongs in a separate commit.

Reply:
:heavy_check_mark: